components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt

```diff
@@ -0,0 +1,183 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.appservices.httpconfig
+
+import com.google.protobuf.ByteString
+import mozilla.appservices.support.RustBuffer
+import mozilla.components.concept.fetch.Client
+import mozilla.components.concept.fetch.MutableHeaders
+import mozilla.components.concept.fetch.Request
+import java.io.InputStream
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+
+/**
+ * Singleton allowing management of the HTTP backend
+ * used by Rust components.
+ */
+object RustHttpConfig {
+    // Protects imp/client
+    private var lock = ReentrantReadWriteLock()
+    @Volatile
+    private var client: Lazy<Client>? = null
+    // IMPORTANT: This must not ever get GCed!!
```

Comment:
Under what circumstances might to get GCed?  In other words, what things are we depending on the calling code not to do, if any?

Reply:
It's more a note for future maintainers of this code. Nasal demons are abound if it gets GCed, so we should be cafeful not to allow that to happen.

Reply:
How do we prevent that? Does the caller need to keep `CallbackImpl` around longer than `RustHttpConfig`, or do anything more?

Reply:
I suggest rewording this comment along the lines of the response above, as an explicit "important note to future maintainers: if you mess around with this code, you have to make sure this cant get GCed".

Reply:
We prevent that by only assigning to this once and never un-assigning it. I'll try and make this comment clearer.

To be clear, there's no danger in the code as I currently wrote it, but is a very large caveat to any future changes to this code -- this is just an extremely important invariant, and otherwise looks like a memory leak / dead store.